### ISSUE TITLE
ClassLoader Leak using TestProfiles (OutOfMemoryError: Metaspace)

### ISSUE THREAD

**Issue by xuckz** [quarkusio/quarkus#38774]
### Describe the bug

Using multiple TestProfiles results in OutOfMemoryError: Metaspace
It seems that running a @QuarkusTest with its own @TestProfile is not cleaned up properly after the Test has finished, and all QuarkusClassLoaders used are permanently stored in MetaSpace. 

Each Test adding its own ClassLoader to MetaSpace until the OutOfMemoryError occurs.

Increasing the MetaSpace for each @QuarkusTest with its own @TestProfile is a limited workaround.

### Expected behavior

The expected behavior would be, that you are not limited to any amount of @TestProfil and that each @TestProfile would not increase the amount of MetaSpace required.

### Actual behavior

Each @QuarkusTest with its own @TestProfile requires the MetaSpace increased.

### How to Reproduce?

_Ready to go repo:
https://github.com/xuckz/quarkus-quickstarts-classloader-leak-demo
https://github.com/xuckz/quarkus-quickstarts-classloader-leak-demo/commit/ccfa14470a8e540336ca086768546cb7cdfe53f1_ 

I used the quarkus-quickstarts/hibernate-orm-panache-quickstart as showcase:

- Checkout https://github.com/quarkusio/quarkus-quickstarts
- Open hibernate-orm-panache-quickstart/pom.xml
- Limit MetaSpace (Since the amount of classes loaded by the quickstart examples is very low, we will have to cap the MetaSpace manually, otherwise we would need a lot of more tests)

```
<plugin>
    <!-- you need this specific version to integrate with the other build helpers -->
    <artifactId>maven-surefire-plugin</artifactId>
    <version>${surefire-plugin.version}</version>
    <configuration>
        <argLine>-XX:MetaspaceSize=128M -XX:MaxMetaspaceSize=256M</argLine>
        <systemPropertyVariables>
            <java.util.logging.manager>org.jboss.logmanager.LogManager</java.util.logging.manager>
            <maven.home>${maven.home}</maven.home>
        </systemPropertyVariables>
    </configuration>
</plugin>
```

- Add 1 empty Test class

```
package org.acme.hibernate.orm.panache.test;

import org.junit.jupiter.api.Test;

import io.quarkus.test.junit.QuarkusTest;
import io.quarkus.test.junit.QuarkusTestProfile;
import io.quarkus.test.junit.TestProfile;

@QuarkusTest
@TestProfile(LeakHunt1Test.LeakHuntTestProfile.class)
public class LeakHunt1Test {

    public static class LeakHuntTestProfile implements QuarkusTestProfile {
    }

    @Test
    void test() {
        System.out.println("asdfasdf");
    }
}
```

- Run mvn test -> Everything runs fine
- Copy LeakHunt1Test.java 9 more times
- Run mvn test -> MetaSpace error

### Output of `uname -a` or `ver`

Linux xxx 5.4.0-167-generic #184-Ubuntu SMP Tue Oct 31 09:21:49 UTC 2023 x86_64 x86_64 x86_64 GNU/Linux

### Output of `java -version`

openjdk version "17.0.9" 2023-10-17 OpenJDK Runtime Environment (build 17.0.9+9-Ubuntu-120.04) OpenJDK 64-Bit Server VM (build 17.0.9+9-Ubuntu-120.04, mixed mode, sharing)

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Apache Maven 3.6.3 Maven home: /usr/share/maven Java version: 17.0.9, vendor: Private Build, runtime: /usr/lib/jvm/java-17-openjdk-amd64 Default locale: en_US, platform encoding: UTF-8 OS name: "linux", version: "5.4.0-167-generic", arch: "amd64", family: "unix"

**Comment by Adelrisk** [comment]
I don't want to be a "I'm having this problem too" commentator, but ... I'm having this problem too. But with only three profiles, each applied to a different test case/test class.

Just for an idea of the scope: real-world project, kotlin and gradle, two entities, few endpoints, 23 passing tests and good coverage (this service is very CRUD-like). (A useful utility microservice that does a single, small job well.)

I then introduced three test profiles that configure two properties differently, such as:

```kotlin
class NoSsiIssuanceV1Profile: QuarkusTestProfile {
    override fun getConfigOverrides(): Map<String, String> {
        return mapOf(
            "quarkus.rest-client.ssi-issuance-v1-api.url" to "",
            "quarkus.rest-client.ssi-issuance-v2-api.url" to "http://ssi-v1:8080"
        )
    }
}
```

I use each profile once to verify the behaviour of injection I programmed into the application, such as:

```kotlin
@QuarkusTest
@TestProfile(NoSsiIssuanceProfile::class)
open class NoSsiClientTest {

    @Inject
    lateinit var sut: Set<ApiVersion>
    @Inject
    lateinit var v1: Optional<SsiIssuanceV1HttpClient>
    @Inject
    lateinit var v2: Optional<SsiIssuanceV2HttpClient>

    @Test
    fun hasCorrectApiVersions() {
        assertThat(sut, equalTo(setOf()))
    }
    @Test
    fun hasNoV1Client() {
        assertThat(v1.isEmpty, equalTo(true))
    }
    @Test
    fun hasNoV2Client() {
        assertThat(v2.isEmpty, equalTo(true))
    }
}
```

When I run all my tests, the JVM prematurely aborts the execution of all three test classes marked with `@TestProfile`. As an experiment, I set a filter so only 10 other tests were executed, in which case "only" one test class with a test profile aborted due to OOM.

This gives the impression that the complexity threshold is quite low before this bug leads to an `OutOfMemoryError`.

**Comment by geoand** [comment]
Can you attach your sample application that leads to the OOME?

**Comment by janek64** [comment]
I am facing the same issue with only two test profiles. All tests of the first profile are running as expected, but the startup of the second profile fails with an `OutOfMemoryError`.

A new observation that I can add is that the issue only occurs since I added `io.quarkus:quarkus-jacoco` to my project. When I remove the dependency, all tests are running as expected.

**Comment by billapepper** [comment]
Another "me too", but I recently started migrating from 3.2.x to 3.8.x (as the LTS is expiring on 3.2.x in July).

I ran into a problem where `@InjectSpy` began throwing a "Please don't pass mock here. Spy is not allowed on mock." error when trying to spy using a smallrye config... this seemingly stems from the Mockito upgrade... but I couldn't find a workaround, so instead, I reworked my application to use `@TestProfile` and override the configuration for the specific tests. Very basic profiles, in a similar fashion to how https://github.com/quarkusio/quarkus/issues/38774#issuecomment-1999958050 did:

```
    public static class PublicPrivateDomains implements QuarkusTestProfile {


        @Override
        public Map<String, String> getConfigOverrides() {
            return Map.of(
                    "prime.landing-service.public-domains", ".domain.com",
                    "prime.landing-service.private-domains", ".domain.net"
            );
        }
    }
```

However, since doing this, the build with several of these tests fails with a OOM error. I have tried increasing the memory using `<argLine>-XX:MetaspaceSize=256M -XX:MaxMetaspaceSize=512M -XX:MaxRAMPercentage=75</argLine>` on the maven-surefire-plugin, but it still runs out of memory in our GH actions.

These tests ran without issues when they were not using separate `@TestProfile`.

**Comment by aylwyne** [comment]
I'm wondering if there is any traction on this issue.  This is blocking our ability to upgrade to 3.8.4 because our test suite uses test profiles quite a lot and we haven't been able to get any alternatives working (see previous comment from billapepper).  We are currently on the previous LTS version (3.2.x) which is expiring soon yet are blocked on upgrading to the 3.8.x LTS release.

**Comment by geoand** [comment]
#41156 might be related and we have a couple of sample changes that may improve things.

**Comment by zilinjak** [comment]
Hi, 
any updates on this ?

**Comment by schonbachler** [comment]
Hi Quarkus Team,

I would like to check if there are any updates or planned milestones regarding this issue.

For our project, this ClassLoader leak has become a major blocker for stable development. As our test suite grows and utilizes different `@TestProfile` configurations, we are increasingly hitting OutOfMemoryError: Metaspace limits in our CI/CD pipelines.

Since stable testing is a core requirement for enterprise-grade Quarkus applications, could you please provide an assessment of the current status? Is there a technical path forward, or can we provide any specific details to help pinpoint the remaining references?

Thank you for your hard work on this!

**Comment by geoand** [comment]
Would you care to try Quarkus from `main`? Asking because @gsmet fixed another CL leak. 
If the problem persists, providing us with a memory dump when the OOME occurs would be helpful

**Comment by cescoffier** [comment]
🤖 Automatically associated with the **WG - Test classloading** Working Group based on AI classification.

**Comment by gsmet** [comment]
I see there is a reproducer above but it might not be all the same issue. 

I think all the latest fixes are in 3.31.4 so if you can reproduce with this version, your issue is not fixed. 

I don’t recommend sharing heap dumps publically but if you’re comfortable sharing one with me, then you can contact me at gsmet at redhat dot com, and I will keep it private and drop it after investigation. 

Another option is to do a call. 

I’m off today but will be back tomorrow.

**Comment by xuckz** [comment]
I merged the main branch of _quarkus-quickstarts_ into my _reproducer_, udating it to **Quarkus 3.32.1.**

I did not spend a lot of time investigating this time, but **the problem still persists**, nothing (at a first glance) changed.

What could i provide to help fix this issue? Running the reproducer is not really hard, i specifically based it on the official quarkus-quickstarts repository, adding only some empty classes.

**Comment by gsmet** [comment]
I'll have a look.

**Comment by gsmet** [comment]
OK, so this is not a simple issue.

To give you some history, I think we fixed most of the class loader leaks that were affecting people in the initial report BUT since then we completely overhauled the test infrastructure to fix some issues... and significantly affect metaspace usage (but also heap space usage).

From what I have seen there are two different issues:

- the fact that we provision all applications upfront require quite a lot of heap space/metaspace before even executing the tests. In your initial setup, you didn't give enough metaspace to even start the tests themselves. And you already configured it quite high.
- once we have sufficient metaspace to actually run the tests, the situation degrades further as we are not dropping the classloaders we don't need anymore and as tests run we populate them even further, thus requiring even more metaspace.

The first problem is by design, even if very unfortunate: I think we need to start thinking about how we can reduce the problem.

The second, from my discussion with @holly-cummins is more considered a bug. But from what I have seen the class loaders are kept around by JUnit infrastructure so I'm not sure we will easily be able to get rid of them... One thing we could do maybe is to close and clear them completely, given I don't think we will need to load more classes after the application/class loader is not useful anymore.

I pushed a couple of pull requests out (one is more or less ready, the other is really just to start a discussion as the implementation is nasty): https://github.com/quarkusio/quarkus/pull/52811 and https://github.com/quarkusio/quarkus/pull/52813 . The first one will fix (again) a class loader leak on Java 25, the second will improve the heap space consumption but NOT the metaspace.

I wonder if we should start revisiting an old PR of mine: https://github.com/quarkusio/quarkus/pull/41608 - which I didn't pursue as we hit some roadblocks.

I prepared an isolated reproducer based on @xuckz 's work: [test-memory-leaks-reproducer.zip](https://github.com/user-attachments/files/25693285/test-memory-leaks-reproducer.zip)

@holly-cummins I would be interested in us having a discussion once you have experimented a bit with the reproducer.

**Comment by holly-cummins** [comment]
Here's what I've got so far. 

I revived https://github.com/quarkusio/quarkus/pull/48830, which I think maybe was closed in error[1]. With it, we go from not closing any of the curated applications (argh!) to closing them when we're done with them (phew!). And the effect is ... negligible. I get, at most, one more test execution before OOMing in metaspace. 

So I added extra code to also aggressively null out the links from the runtime classloader to the curated application, and close the runtime classloader when we're done with it. Still no effect. 

I looked in the heap dump, and there's still 36 classloaders. This was surprising to me. At least half of them seem to be retained by threads, on which they were set as the TCCL (and it might be almost all of them, actually). So I might be able to shift something by being more aggressive in unsetting TCCLs. Most of the threads which are holding on to a classloader seem to be associated with testcontainers. I think it's something in the area of dev services, because if I disable dev services and use an external database, the tests run through without OOM.

[1] It couldn't get merged, because it broke all sorts of things in CI, but the issue was a real one.

**Comment by holly-cummins** [comment]
... and the latest update. 

- We're not closing curated applications, which is a bug. https://github.com/quarkusio/quarkus/pull/48830 fixes it, but introduces/exposes other problems. I will try and revive it. 
- However, closing the curated applications makes very little difference. Even when we close the applications, we OOM and have 36 QuarkusClassLoaders on the heap. 
- Many of these classloaders are referenced by TCCLs. It looks like TestContainers takes the active TCCL when containers are created or started and sets it as a TCCL on `testcontainers-wait` and `testcontainers-ryuk` threads. I think you'd looked into this already, @gsmet. I hacked around to prevent this, and it didn't improve much, because the affected loaders were also being held elsewhere.
- Most of the 36 loaders are retained by JUnit `ClassSelectors`, which hold a reference to the test class until the end of the run. I thought this would only hold a reference to the runtime loader, but the runtime loader has a parent, and that parent is the base runtime loader. That holds a reference to an application, which holds a reference to a deployment loader. So that's 3 quarkus classloaders per test class. I can remove the references to the application and deployment classloader on close and that gets me down to 31 classloaders, but the OOM behaviour is exactly the same. 
- The FacadeClassLoader has one or two extra classloaders, and I shut down the FCL aggressively early, so those classloaders could close before test execution. It changed nothing. 
- In my heap dumps, I'm sometimes seeing a 50-50 mix of closed and open applications. I'm a bit puzzled by this, but what I think is happening is that a lot of the shutdown logic is happening on other threads, and these tests are so fast that the shutdown of previous applications can be lagging behind. However, slowing the tests down with a sleep didn't change the OOM behaviour. I also experimented with not shutting down dev services, just to get app shutdown faster, but that also didn't change much. 
- As mentioned, if I don't use dev services and use an external db, the tests pass with no problems, despite the 36 classloaders.
- I noticed that Testcontainers starts a ryuk reaper for every container. This means that with lots of profiles, we get lots of reapers. (The reapers don't start until the test executes, but then there's a latency in shutting them down.) If I set testcontainers reuse to true, all the duplicated reapers and postgres containers vanish, and ... the OOM behaviour is the same. So if the problem is happening because of the testcontainers infrastructure, it's in artifacts "before" the decision to launch an external container.
- I made testcontainers parent-first, to ensure that all testcontainers statics were truly global, and to ensure we only had one copy of the testcontainers classes in our tree. It changed ... nothing.

So I'm going to:
- Try and get #48830 to a point where closing the curated applications doesn't break lots of stuff (maybe hard, but clearly 'correct')
- Null out the deployment classloader in closed applications (easy)

... but that's not enough to fix the problem. I feel like the 36 classloaders are obviously bad, but they're maybe not the right tree for this particular problem, and there's something in how we handle dev services that might be causing problems. That also suggests that more resource sharing between classloaders, while generally beneficial, might not help here. I'll see if I can hack things to try and isolate the problem more. 

I also suggest we put Guillaume's reproducer into source as a test, since this feels like the kind of thing that will regress easily and often. But we need to get it passing first. :)

**Comment by gsmet** [comment]
> That also suggests that more resource sharing between classloaders, while generally beneficial, might not help here. I'll see if I can hack things to try and isolate the problem more.

From a metaspace POV, I think it makes perfect sense to try to avoid having 9 augmentation class loaders. If we put this leak aside, the amount of metaspace we need to initialize the 9 ``CuratedApplication`` is already far higher than I would expect.

As for sharing resources between the class loaders, it won't help the metaspace story, but it will help the heap space story. At the moment, typically, we have 60+ megabytes of heap used for the numerous copies of `PathTreeClassPathElement`.

**Comment by gsmet** [comment]
> Many of these classloaders are referenced by TCCLs. It looks like TestContainers takes the active TCCL when containers are created or started and sets it as a TCCL on testcontainers-wait and testcontainers-ryuk threads. I think you'd looked into this already, @gsmet. I hacked around to prevent this, and it didn't improve much, because the affected loaders were also being held elsewhere.

Yeah, I had a look at that a long time ago. And the only solution I found was: https://github.com/quarkusio/quarkus/pull/42560 but IIRC it opened a can of worms.

**Comment by gsmet** [comment]
Also, in general, I wouldn't label dropping a reference to the CL as "doesn't help". It does help. We just need to get rid of ALL of them for it to be effective.

**Comment by holly-cummins** [comment]
There may be some more subtle things we can do, too. Classes can be duplicated between the augmentation and deployment, and between runtime and base runtime classloaders. Reducing that duplication would reduce metaspace usage. How do they get duplicated? I think it happens because we sometimes use the 'parent' as a TCCL, and sometimes a child, and if we're not consistent, the same class will be loaded by both. So we might look into our hygiene for that (but unpicking TCCL issues is super-hard).

**Comment by gsmet** [comment]
> Classes can be duplicated between the augmentation and deployment, and between runtime and base runtime classloaders.

I think our main problem atm is that classes are duplicated in our 36 class loaders :). We can refine later and have a look at other issues but I don't think what you mention here is the culprit for the massive amount of metaspace we need.

**Comment by gsmet** [comment]
> I also suggest we put Guillaume's reproducer into source as a test, since this feels like the kind of thing that will regress easily and often. But we need to get it passing first. :)

It was actually @xuckz 's reproducer. I just extracted it for easier consumption.

**Comment by holly-cummins** [comment]
> Also, in general, I wouldn't label dropping a reference to the CL as "doesn't help". It does help. We just need to get rid of ALL of them for it to be effective.

Let me rephrase it: "The results of my experiments were that symptoms stayed the same, and the change didn't allow the tests to progress any further. I was hoping to see an incremental improvement in behaviour from this incremental change, but to my surprise, I did not."

**Comment by holly-cummins** [comment]
> > Classes can be duplicated between the augmentation and deployment, and between runtime and base runtime classloaders.
> 
> I think our main problem atm is that classes are duplicated in our 36 class loaders :). We can refine later and have a look at other issues but I don't think what you mention here is the culprit for the massive amount of metaspace we need.

I just checked, and there isn't actually that much class duplication, at least not between parent and child. The runtime and deployment classloaders hold almost no classes, and it's the augmentation and base runtime where the classes are. And as you say, those are the ones which are most amenable to sharing.

**Comment by gsmet** [comment]
> Let me rephrase it: "The results of my experiments were that symptoms stayed the same, and the change didn't allow the tests to progress any further. I was hoping to see an incremental improvement in behaviour from this incremental change, but to my surprise, I did not."

Yeah, my point was that every time I chased these leaks, I had the impression I wasn't making any progress, until I did. If you have 10 references to the CL, and you reduce it to 1 reference, you still have a leak. You made progress though. But either we find a way to get to 0, or indeed, we won't be able to solve the problem.

**Comment by holly-cummins** [comment]
> > Let me rephrase it: "The results of my experiments were that symptoms stayed the same, and the change didn't allow the tests to progress any further. I was hoping to see an incremental improvement in behaviour from this incremental change, but to my surprise, I did not."
> 
> Yeah, my point was that every time I chased these leaks, I had the impression I wasn't making any progress, until I did. If you have 10 references to the CL, and you reduce it to 1 reference, you still have a leak. You made progress though. But either we find a way to get to 0, or indeed, we won't be able to solve the problem.

Understood, and it's a good point. For clarity, in some of the experiments, I dropped whole classloaders, not references to them. I expected that to give an incremental improvement. But it turns out the classloaders I was dropping were very small, so removing them didn't make a dent. I also established that we can run the tests quite happily with the exact same metaspace and all 36 quarkus classloaders in the heap, so I'd like to understand what's different in that scenario a bit better, since I think it might offer one possible path to a solution. I mean, even if we do consolidate classloaders, we want to understand what the _other_ really expensive thing in the metaspace is, right? That might allow us to drop our metaspace requirements even further.

**Comment by gsmet** [comment]
Just a note that I haven't forgotten about you all. I have created this PR today: https://github.com/quarkusio/quarkus/pull/53656 .

First step will be for it to pass full CI, second step will be for it to pass @holly-cummins' scrutiny :).

**Comment by gsmet** [comment]
I was able to get https://github.com/quarkusio/quarkus/pull/53656 green and it has been quite of a challenge.

It improves the situation and I'm working on some other investigations to see if we can further improve the situation.
